### PR TITLE
475 no method matching for neurotreemodel

### PR DESCRIPTION
**Pull request by pat-alt**
@abuszydlik your hunch was right and this should do it now. 

Not the most elegant way to infer the element type of the underlying `Flux.Chain` but not really sure how else to address this. 

@jeremiedb how are you doing? Hope the remainder of your Eurotrip was fun! Is there a better way to infer the element type of the weights `W` of the underlying `NeuroTree{W, ...}`? I'll merge this as soon as tests pass, so @abuszydlik can continue his work, but would much appreciate your thoughts if any. Thanks!

### CODE REVIEW

**Comment by jeremiedb** [comment]
Hi Patrick! Thanks I really enjoyed discoevering NL! 

Regarding the issue with the inference type, NeuroTrees enforces by design the layers to be Float32. 
So assuming that the case where X is vector is first rehsape into a Matrix, it would be safe to use something like:
```julia
X = convert(Matrix{Float32}, X[:, :]);
```

That being said, I'd be curious about how the `Models.logits` are  `Models.probs` are used in the package. Ie. is it for a single shot inference of the models, or is it to be used throught iterative loops while training another model? I'm asking as for background, NeuroTreeModels was built from a Table/DataFrame first perspective, handling behind the scene the types for features matrix and target vector, and ensuring that at inference proper variable names were present. Hence, calling the model on a matrix `M.fitresult(X)` wasn't intended as a user facing method, as opposed as calling it over a Table/DataFrame, which handles the creation of the iterator and loop over batches to accumulate the predictions over a potentially too large dataset to fit into a single forward pass of the model. 
In short, if it's expected that X can be a large dataset, it may pose a challenge if the NeuroTree model is very large (just like any other large Flux model). And if for quick inference on batch size like data, then workflow wise, it may be advise to ensure that the full feature matrix is first converted into Float32 before going into iterations.

**Comment by pat-alt** [comment]
Glad to hear @jeremiedb :)

Thanks for clarifying! 

> That being said, I'd be curious about how the Models.logits are Models.probs are used in the package. Ie. is it for a single shot inference of the models, or is it to be used throught iterative loops while training another model?

For gradient-based counterfactual generators, `Models.logits` is called iteratively as we take gradient steps in the feature space, but not to [train](https://github.com/JuliaTrustworthyAI/CounterfactualExplanations.jl/blob/65e38f380ff4b1c15966b342eb7e17de210dfee4/ext/NeuroTreeExt/neurotree.jl#L51) the NeuroTreeModels or any derivative model.

> In short, if it's expected that X can be a large dataset, it may pose a challenge if the NeuroTree model is very large (just like any other large Flux model). And if for quick inference on batch size like data, then workflow wise, it may be advise to ensure that the full feature matrix is first converted into Float32 before going into iterations.

That's good to know, thank you. I suppose that for large datasets this may then cause performance bottlenecks. We do have a [method](https://github.com/JuliaTrustworthyAI/CounterfactualExplanations.jl/blob/2bf113a3531277f80bec654e8df2161063e456f6/src/data_preprocessing/counterfactual_data.jl#L200) to wrap Tables, but that still then transforms the Table into a Matrix for downstream tasks. 

The main reason we've designed things this way is that gradient-based counterfactual search involves gradient computations with respect to features (as opposed to model parameters) and matrices are easier to handle. 

So I think that to change this we'll have to revisit some core design choices. For now I think I'm happy with the conversion [here](https://github.com/JuliaTrustworthyAI/CounterfactualExplanations.jl/blob/65e38f380ff4b1c15966b342eb7e17de210dfee4/ext/NeuroTreeExt/neurotree.jl#L86), although I might simplify it further since we can just generally assume Float32 as the element type. 

Thanks for your advise!
